vulcan/lib/commands.rb

```diff
@@ -18,6 +19,174 @@ def setup(config)
     end
   end
 
+  class Workflow
+    include Etna::CommandExecutor
+
+    class Run
+      def execute(project_name, workflow_name)
+        @project_name = project_name
+        @workflow_name = workflow_name
+
+        while true
+          orchestration.run_until_done!(storage)
+          find_all_output_ui_queries.each do |query_name, source, path|
+            if self.respond_to?(query_name.to_sym)
+              self.send(query_name.to_sym, source, path)
+            else
+              show_data(source, path)
+            end
+          end
+
+          while (input = find_next_primary_input)
```

Comment:
Primary inputs, like primary outputs, are definitely the special case of having types but no run information encoded.  Perhaps we could provide type information and use that as the determinate of gathering inputs, but I still have skepticism if that is wise.  In any case, for primary inputs, we definitely only have the type to go off of.